### PR TITLE
[macOS] Bring up "hello_world" devicelab, compilation test for x86.

### LINKED ISSUE #110079: [macOS] Bring up "hello_world" devicelab, compilation test for x86.

### PR DESCRIPTION
**Pull request by a-wallen**
#### Related
fixes https://github.com/flutter/flutter/issues/110079

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: .ci.yaml
Comment (a-wallen) [inline .ci.yaml:2552]: It would be nice if we could run this on both M1 and Intel Macs. I wasn't sure how to configure that in a target, so this is my best effort.

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (cbracken) [inline dev/devicelab/lib/tasks/perf_tests.dart:1493]: Use [`File.statSync`](https://api.flutter.dev/flutter/dart-io/FileSystemEntity/statSync.html) here, then grab the [`size`](https://api.flutter.dev/flutter/dart-io/FileStat/size.html) from that.

File: .ci.yaml
Comment (keyonghan) [reply .ci.yaml:2552]: This enables task in `x64` only.
The arm64 support is tracked here: https://github.com/flutter/flutter/issues/104034
F.Y.I. https://github.com/flutter/flutter/pull/109889 is attempting to run host only task in either arm. /cc @jmagman

**Review by yaakovschectman — APPROVED** [review]
LGTM plus Chris' comment

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (jmagman) [inline dev/devicelab/lib/tasks/perf_tests.dart:1480]: Rename this
```suggestion
            _findAppInBuildDirectory(buildDirectoryPath);
```

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (jmagman) [inline dev/devicelab/lib/tasks/perf_tests.dart:1495]: This is double the size of a downloaded app, because it contains both arm64 and x86.  That's the same reason why `getSizesFromIosApp` thins the app first.

Also, `app_framework_uncompressed_bytes` and `flutter_framework_uncompressed_bytes` will be better to report in order to match iOS, otherwise you are including things like the example app icon size, which doesn't matter.

I'd prefer if iOS and macOS looked exactly the same, except for the logic that actually finds the `.app` directory.  The more we can combine iOS and macOS logic in the tests and tool, the better.

**Review by jmagman — DISMISSED** [review]

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (cbracken) [inline dev/devicelab/lib/tasks/perf_tests.dart:1495]: It looks like we're producing a tar archive just to measure the bytes. Since tar archives are basically just a concatenation of the files in the archive plus a header before each one with the path/size/permission/timestamp data, as well as some zero-padding, this is roughly equivalent to just summing up the file sizes collected while iterating over the files. I'd just do that instead.

A followup thought -- unlike iOS, there are multiple ways developers can distribute a macOS app: via the app store or individually, typically in a compressed disk image.

We should probably decide how we want to measure this. One of:
1. The sum of the components of the files in the `.app` bundle. Current implementation.
2. The size of a distributable that would be uploaded to the app store (i.e. the equivalent of an IPA).
3. The size of the compressed app bundle (e.g. maybe a zip file).

Personally, option 1 or 2 seem most reasonable. The iOS code above is a bit weird; it's generating a compressed tar archive. I thought IPAs were a zipfile though, but I suppose a gzipped tar is 'close enough'. 

Maybe we do that for consistency?

Comment (a-wallen) [reply dev/devicelab/lib/tasks/perf_tests.dart:1495]: cc @jmagman

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (a-wallen) [reply dev/devicelab/lib/tasks/perf_tests.dart:1480]: Before I make the change - I want to provide some additional context so that we are on the same page.

This is the logic in the iOS branch of the switch case for `_compileTest`.
https://github.com/flutter/flutter/blob/a6c2e711e455bed6948f4f57620840e71c3d9eba/dev/devicelab/lib/tasks/perf_tests.dart#L1411-L1416

I copied it over to the macOS branch of the switch case, but I noticed [`_findIosAppInBuildDirectory`](https://github.com/flutter/flutter/blob/a6c2e711e455bed6948f4f57620840e71c3d9eba/dev/devicelab/lib/tasks/perf_tests.dart#L1946-L1953) was being used elsewhere to grab the `*.app` directory for iOS.

https://github.com/flutter/flutter/blob/a6c2e711e455bed6948f4f57620840e71c3d9eba/dev/devicelab/lib/tasks/perf_tests.dart#L1946-L1953

If I use the function call, I will rename to `_findAppInBuildDirectory(buildDirectoryPath)` since it will be used by`iOS` and `macOS`. Let me know what you think about that.

While I'm at it, should I refactor this: https://github.com/flutter/flutter/blob/a6c2e711e455bed6948f4f57620840e71c3d9eba/dev/devicelab/lib/tasks/perf_tests.dart#L1411-L1416

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (a-wallen) [reply dev/devicelab/lib/tasks/perf_tests.dart:1495]: https://github.com/flutter/flutter/blob/a6c2e711e455bed6948f4f57620840e71c3d9eba/dev/devicelab/lib/tasks/perf_tests.dart#L1423-L1426

`getSizedFromIosApp` does not appear to thin the app first before assigning a `releaseSizeInBytes`. Am I mistaken?

Comment (jmagman) [reply dev/devicelab/lib/tasks/perf_tests.dart:1495]: You're right.  I guess the point of that one (since it's an ipa) was to see how big it is to submit to the App Store.  And for iOS that only means arm64 since we recently stopped supporting 32-bit armv7, so the "thinning" isn't even doing anything, so that can be removed.

For macOS we could just build `arm64` and then skip the thinning (done below by setting `FLUTTER_XCODE_ARCHS`, which should get parsed and passed into Xcode).  Or we could do as @cbracken pointed out and just not bother with it and just get the full app size with both archs.

How about something like this (though I didn't run it):
```dart
    switch (deviceOperatingSystem) {
      case DeviceOperatingSystem.ios:
      case DeviceOperatingSystem.macos:
        options.insert(0, deviceOperatingSystem == DeviceOperatingSystem.ios ? 'ios' : 'macos'); // Enhanced enum instead?
        options.add('--tree-shake-icons');
        options.add('--split-debug-info=infos/');
        watch.start();
        await flutter('build', options: options, environment: <String, String>{
          'FLUTTER_XCODE_ARCHS': 'arm64',
        });
        watch.stop();
        final Directory appBuildDirectory = Directory(path.join(cwd, 'build')); // This should be safe since there's a clean right before it.
        final Directory? appBundle = appBuildDirectory
            .listSync()
            .whereType<Directory?>()
            .singleWhere((Directory? directory) =>
              directory != null && path.extension(directory.path) == '.app',
              orElse: () => null);
        if (appBundle == null) {
          throw 'Failed to find app bundle in ${appBuildDirectory.path}';
        }
        final String appPath =  appBundle.path;
        releaseSizeInBytes = ... // Don't `tar`, zip or sum up the files as Chris suggests in his comment.
        if (reportPackageContentSizes) {
          metrics.addAll(await getSizesFromDarwinApp(appPath, deviceOperatingSystem));
        }
        break;
```
```dart
  static Future<Map<String, dynamic>> getSizesFromDarwinApp(String appPath, DeviceOperatingSystem operatingSystem) async {
    final String frameworkDirectory;
    if (operatingSystem == DeviceOperatingSystem.ios) {
      frameworkDirectory = path.join(appPath, 'Frameworks');
    } else {
      // macOS
      frameworkDirectory = path.join(appPath, 'Contents', 'Frameworks');
    }
    final File appFramework = File(path.join(frameworkDirectory, 'App.framework', 'App'));
    final File flutterFramework;

    if (operatingSystem == DeviceOperatingSystem.ios) {
      flutterFramework = File(path.join(frameworkDirectory, 'Flutter.framework', 'Flutter'));
    } else {
      // macOS
      flutterFramework = File(path.join(frameworkDirectory, 'FlutterMacOS.framework', 'FlutterMacOS'));
    }

    return <String, dynamic>{
      'app_framework_uncompressed_bytes': await appFramework.length(),
      'flutter_framework_uncompressed_bytes': await flutterFramework.length(),
    };
  }
```

File: .ci.yaml
Comment (keyonghan) [inline .ci.yaml:2561]: @jmagman has been driving the effort to categorize different tests: host-only, devices, etc. to better use our bot resources. Let's land the infra change first so you can use directly: https://github.com/flutter/flutter/pull/110050

**Comment by a-wallen** [comment]
Blocked until https://github.com/flutter/flutter/pull/110050 lands

File: .ci.yaml
Comment (a-wallen) [reply .ci.yaml:2561]: Got it 👍

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (a-wallen) [reply dev/devicelab/lib/tasks/perf_tests.dart]: @jmagman comment is outdated due to the following change
https://github.com/flutter/flutter/blob/d70df45883b810910e8b90a9e5aec1f750869b7b/dev/devicelab/lib/tasks/perf_tests.dart#L1960-L1968

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (a-wallen) [reply dev/devicelab/lib/tasks/perf_tests.dart]: @cbracken instead of archiving the compilation output. I've chosen record the size of the build output in the number of bytes. Please resolve as you see fit:
https://github.com/flutter/flutter/blob/d70df45883b810910e8b90a9e5aec1f750869b7b/dev/devicelab/lib/tasks/perf_tests.dart#L1465-L1471

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (a-wallen) [reply dev/devicelab/lib/tasks/perf_tests.dart]: @cbracken at the time of writing, **Option 1** is the approach that has been taken. Please comment/resolve as you see fit:
https://github.com/flutter/flutter/blob/d70df45883b810910e8b90a9e5aec1f750869b7b/dev/devicelab/lib/tasks/perf_tests.dart#L1437-L1479

File: .ci.yaml
Comment (a-wallen) [reply .ci.yaml]: @keyonghan I have changed this PRs `.ci.yaml` tests to use a `Mac_arm64` target. 
https://github.com/flutter/flutter/blob/d70df45883b810910e8b90a9e5aec1f750869b7b/.ci.yaml#L2561-L2568
This PR will be blocked until (https://github.com/flutter/flutter/pull/110050) lands, but we will not accumulate any technical debt.

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (a-wallen) [reply dev/devicelab/lib/tasks/perf_tests.dart]: @jmagman this following code should address the feedback above
https://github.com/flutter/flutter/blob/d70df45883b810910e8b90a9e5aec1f750869b7b/dev/devicelab/lib/tasks/perf_tests.dart#L1439-L1479

Thank you for the feedback 👍

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (jmagman) [inline dev/devicelab/lib/tasks/perf_tests.dart:1437]: Instead of moving the iOS logic down, instead move the `case DeviceOperatingSystem.macos:` logic up to make git spelunking easier.

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (jmagman) [inline dev/devicelab/lib/tasks/perf_tests.dart:1441]:
```suggestion
        if (deviceOperatingSystem == DeviceOperatingSystem.ios) {
```

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (jmagman) [inline dev/devicelab/lib/tasks/perf_tests.dart:1465]: By no longer compressing, this will increase the iOS size benchmark trend.  The engine sheriff may contact you about this.

@zanderso

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (jmagman) [inline dev/devicelab/lib/tasks/perf_tests.dart:1473]: I know the rest of this file uses `dynamic`, but you should avoid `dynamic` in null safe code.  Use `Object` or `Object`.

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (jmagman) [inline dev/devicelab/lib/tasks/perf_tests.dart:1560]:
```suggestion
    if (!operatingSystem.isDarwin) {
```

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (jmagman) [inline dev/devicelab/lib/tasks/perf_tests.dart:1585]: I know you copied this from me but you could have just one if and set `frameworkDirectory` and `flutterFramework` at the same time.

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (jmagman) [inline dev/devicelab/lib/tasks/perf_tests.dart:1596]: Ha, the 😞 was for your benefit, probably shouldn't leave it as a comment since it causes more questions than answers.

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (a-wallen) [reply dev/devicelab/lib/tasks/perf_tests.dart:1437]: Does it not matter that the switch cases are alphabetized?

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (a-wallen) [reply dev/devicelab/lib/tasks/perf_tests.dart:1585]: With this way, we use one conditional tree for each assignment. You are suggesting that we use one tree for all assignments. I personally think that this way is more readable because declaration and assignment logic are contained to one region of the file. More importantly, that region is not mixed with other assignments. What do you think about that?

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (zanderso) [reply dev/devicelab/lib/tasks/perf_tests.dart:1465]: Please retain the compression step for iOS.

File: .ci.yaml
Comment (keyonghan) [inline .ci.yaml:2561]: Sorry for the back and forth here. Host only targets on arm64 is now blocked on issues in https://github.com/flutter/flutter/pull/109889.
To unblock, you may want to run it in `Mac_arm64_ios` platform.

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (jmagman) [reply dev/devicelab/lib/tasks/perf_tests.dart:1437]: Nah.

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (jmagman) [reply dev/devicelab/lib/tasks/perf_tests.dart:1465]: @zanderso Is this metric used for something I'm not aware of?

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (jmagman) [reply dev/devicelab/lib/tasks/perf_tests.dart:1585]: I was thinking this, but I don't really have a strong preference.
```dart
    late final String frameworkDirectory;
    late final File flutterFramework;
    if (operatingSystem == DeviceOperatingSystem.ios) {
      frameworkDirectory = path.join(
        appPath,
        'Frameworks',
      );
      flutterFramework = File(path.join(
        frameworkDirectory,
        'Flutter.framework',
        'Flutter'
      ));
    } else if (operatingSystem == DeviceOperatingSystem.macos) {
      frameworkDirectory = path.join(
        appPath,
        'Contents',
        'Frameworks',
      );
      flutterFramework = File(path.join(
        frameworkDirectory,
        'FlutterMacOS.framework',
        'FlutterMacOS'
      ));
    }

    final File appFramework = File(path.join(
      frameworkDirectory,
      'App.framework',
      'App',
    ));
```

File: .ci.yaml
Comment (jmagman) [reply .ci.yaml]: @a-wallen how about leaving off the arm test, and filing a new issue to add it once https://github.com/flutter/flutter/pull/110050  or similar lands?

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (a-wallen) [reply dev/devicelab/lib/tasks/perf_tests.dart]: It's not exactly what you are suggesting, but I think this captures your suggestion as well as @gspencergoog 
https://github.com/flutter/flutter/blob/000cd8bb89ee4abf9f7fc49163a76a7cca74712d/dev/devicelab/lib/tasks/perf_tests.dart#L1560-L1593

File: .ci.yaml
Comment (a-wallen) [reply .ci.yaml]: That's probably what is best now that https://github.com/flutter/flutter/pull/109889 seems to be blocked.

Comment (a-wallen) [reply .ci.yaml]: @keyonghan I wrote a handful of draft PRs with `Mac_arm64` as a target. Do you have an idea of when `Mac_arm64` might be available?

File: .ci.yaml
Comment (a-wallen) [reply .ci.yaml:2561]: https://github.com/flutter/flutter/pull/109891#discussion_r952032899

Comment (a-wallen) [reply .ci.yaml:2561]: https://github.com/flutter/flutter/pull/109891#discussion_r953065815

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (zanderso) [reply dev/devicelab/lib/tasks/perf_tests.dart:1430]: We track release size in SkiaPerf https://flutter-flutter-perf.skia.org/e/?queries=sub_result%3Drelease_size_bytes%26test%3Dhello_world_ios__compile

Comment (jmagman) [reply dev/devicelab/lib/tasks/perf_tests.dart:1430]: Oh for sure, I've looked at that data. Is it a problem to rebenchmark it to make it consistent between macOS and iOS? The more we can make the darwin tests and benchmarks the same between iOS and macOS, the less maintenance burden it is since improvements to one platform make it to the other.  It also makes it more meaningful to compare benchmarks across platforms.  We could compress the macOS app but that doesn't really mean anything.  We could also make an iOS specific code path to compress it, but I don't care so much whether we're measuring it compressed or not, as long as we're keep tracking of it growing or shrinking over time.

Comment (zanderso) [reply dev/devicelab/lib/tasks/perf_tests.dart:1430]: Yes, we do want to track the compressed size. An example of a reason to track both compressed and uncompressed size is that a change to the Dart snapshot format might make the uncompressed size smaller while making the compressed size bigger due to compressing less well as a result of a formatting or data layout change.

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (loic-sharma) [inline dev/devicelab/lib/tasks/perf_tests.dart:1562]:
```suggestion
    switch (deviceOperatingSystem) {
```

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (loic-sharma) [inline dev/devicelab/lib/tasks/perf_tests.dart:1571]:
```suggestion
          'Flutter',
```

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (loic-sharma) [inline dev/devicelab/lib/tasks/perf_tests.dart:1583]:
```suggestion
          'FlutterMacOS',
```

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (jmagman) [reply dev/devicelab/lib/tasks/perf_tests.dart:1430]: Does it make sense to do the same for macOS then?

**Comment by a-wallen** [comment]
Now blocked on https://github.com/flutter/flutter/pull/109889
- [ ] https://github.com/flutter/flutter/pull/109889

**Comment by a-wallen** [comment]
@jmagman I'm going to drop bc63b3b0ae and make a todo in https://github.com/flutter/flutter/issues/110126.

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (a-wallen) [reply dev/devicelab/lib/tasks/perf_tests.dart]: @zanderso thanks for letting us know. 

This is the result from running `_compileApp` on `iOS` and `macOS` for this patch. I think it's what we need. 

https://github.com/flutter/flutter/blob/5c1b21c4f97190a9b3144cda029cc626ffa03ee8/dev/devicelab/lib/tasks/perf_tests.dart#L1475-L1478

https://github.com/flutter/flutter/blob/5c1b21c4f97190a9b3144cda029cc626ffa03ee8/dev/devicelab/lib/tasks/perf_tests.dart#L1572-L1575

Comment (zanderso) [reply dev/devicelab/lib/tasks/perf_tests.dart]: I think if we're tracking compressed size for one target, then that is sufficient, but it certainly doesn't hurt to collect it for more platforms if that simplifies this code.

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (jmagman) [inline dev/devicelab/lib/tasks/perf_tests.dart:1407]: This comment doesn't make sense, ipa isn't a thing on macOS.  How about:
```dart
        // Validate changes in Dart snapshot format and data layout do not change compression size.
        await exec('tar', <String>['-zcf', 'build/app.tar.gz', appPath]);
        releaseSizeInBytes = await file('$cwd/build/app.tar.gz').length();
```

**Review by jmagman — COMMENTED** [review]
I wasn't done reviewing this, can you make a new PR to follow up?

**Review by cbracken — COMMENTED** [review]
🎉

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (a-wallen) [reply dev/devicelab/lib/tasks/perf_tests.dart:1407]: uh-oh sorry about that!

**Pull request merged**
